### ISSUE TITLE
Make _getFunction method public in t.w.xmlrpc.XMLRPC

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4836]
|<img alt="psykidellic's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| psykidellic reported|
|-|-|
|Trac ID|trac#4836|
|Type|enhancement|
|Created|2011-01-28 21:14:43Z|
|Branch|https://github.com/twisted/twisted/tree/xmlrpc-getfunction-4836|

Sometimes an end application would like to choose the implementation method based on functionName, which mayb or may not be part of the XMLRPC instance.

So one could override the getFunction method in such cases.

Attachments:

* [twisted_4836.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/fd3/fd3035f1d3859400f44bc727d1d64f84d46d4d4a/487613707dd9bbee9aee9e7eaca4f2a102169235.patch) (6255 bytes) - added by psykidellic on 2011-02-01 18:51:03Z - Patch for 4836
* [issue_4836.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/fd3/fd3035f1d3859400f44bc727d1d64f84d46d4d4a/dc3dc5a8b7ef2f1b2f8cbf83c9a605b4b7a14a71.patch) (8585 bytes) - added by psykidellic on 2011-02-10 00:24:24Z - Updated patch

<details><summary>Searchable metadata</summary>

```
trac-id__4836 4836
type__enhancement enhancement
reporter__psykidellic psykidellic
priority__normal normal
milestone__ 
branch__branches_xmlrpc_getfunction_4836 branches/xmlrpc-getfunction-4836
branch_author__exarkun__psykidellic exarkun, psykidellic
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1296249283000000 1296249283000000
changetime__1302980714000000 1302980714000000
version__None None
owner__exarkun exarkun
cc__psykidellic
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="psykidellic's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|psykidellic commented|
|-|-|

Hi

Find attached my patch for the issue. It applies to trunk at rev 30599. 

As discussed with exarkun at #python, I have taken care of 3 topics:

* Code change: Mostly it was find and replace _getFunction to getFunction. The docstring for it already mentioned about overriding. I think that is good enough. If not let me know and I will try to add something to it.

* Test cases: Updated _getFunction call to corresponding getFunction. Added two simple test case to verify that a derived method is called. The change cleanly passes all the 52 old tests and 2 new ones. So now we have total 54 tests for xmlrpc. 

* Documentation: I added an example in web/howto/xmlrpc.xhtml for the same. Its almost duplicated from the previous one so maybe we want to make it succinct. Also, I am not sure what I could write for the heading so I just used: Using your own Function getter. I believe this could be improved. Any thoughts?

Let me know if something is missing or can be improved.

**Comment by twisted-trac** [comment]
|<img alt="taos's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|taos commented|
|-|-|

Suggestion for line 170 of test_xmlrpc.py, rather than writing the dictionary expanded like that you can simply place it all on a single line
```
x = {'echo':self.echo}
```
Seems a lot neater in my opinion, I'm sure the reviewer will have more to add or suggest.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to psykidellic|
|-|-|

 * You need a NEWS fragment
 * test_errors is missing a docstring.
 * There should be 2 blank lines between methods and 3 between classes (see http://twistedmatrix.com/trac/browser/trunk/doc/core/development/policy/coding-standard.xhtml)
 * It's unfortunate that another test with a server is added to test_xmlrpc. I won't blame you for that, but you really don't need a listenTCP to test this change.
 * Overriding getFunction has the side effect of breaking listMethods. While we shouldn't enforce the contract I think:
  * We should rename `_listFunctions` to `listFunctions`
  * Indicate in `getFunction` that overriding it means that overriding `listFunctions` as well would be a good idea, or otherwise introspection is broken.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="psykidellic's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|psykidellic commented|
|-|-|

Thanks for the detailed review. Find attached an updated patch which fixes the above mentioned points.

*  Added NEWS fragment - 4836.feature

*  Added docstrings to test_errors.

*  Updates all with correct blank lines.

*  Well, I was not sure how to do it otherwise? The older testcases has its own XMLRPC instances which I didnt really want to muddle with. If you have any suggestion to do it correctly - I will update this patch again.

*  Updated relevant docs/method names for listFunctions. Added testcase to handle the change too.


Let me know if the patch can be still be improved.

**Comment by twisted-trac** [comment]
|<img alt="psykidellic's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|psykidellic removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph set owner to psykidellic|
|-|-|

Thanks again for your updated patch.  There are still a couple of issues with it, but I think it's pretty close:

  1. The naming here should be more consistent.  For example, the docstring for `listFunctions` says it lists the names of XMLRPC methods.  I realize xmlrpc itself is somewhat inconsistent between 'procedure' and 'method', but it never uses the term 'function'.  We should pick one term and stick with it; I suggest "procedure" as that is used less frequently elsewhere.  (Also, as long as we're talking about naming conventions, I don't like 'get' as a prefix; it should be `lookupProcedure`.)  I'll use the modified terminology for the rest of these points, but if you like "method" better you can feel free to use that in place of "procedure".
  1. as per the [coding standard](http://twistedmatrix.com/documents/10.2.0/core/development/policy/coding-standard.html#auto11), `lookupProcedure` and `listProcedures` should both have `@since` declarations.  11.0 will be our next release.
  1. The `TestLookupProcedure` test helper should have docstrings.
  1. The coding standard suggests wrapping at 80 characters.  This looks like it's wrapped somewhere closer to 50; it looks a little odd.
  1. The test docstring style that we now prefer is declarative and specific.  For example, your docstring that says "... should prevent introspection from being broken" is vague.  Don't use words like "broken" or "working" or "correctly" or "incorrectly".  Explain what non-broken introspection behavior is.
  1. Don't use a real reactor if you can possibly avoid it.  For example, your `tearDown` is oversimplified and may sometimes fail. [See Jonathan Lange's post on how hard this is for more details.](http://blackjml.livejournal.com/23029.html)  `twisted.web.xmlrpc`'s reactor integration is pretty light, so you can fix these tests pretty easily by giving `Proxy` a `reactor` attribute and using `twisted.test.proto_helpers.MemoryReactor` in the tests, rather than returning a `Deferred`.
  1. The code example should use `class="py-listing"` rather than `<pre>`.  Keeping python files external makes them easier to edit, easier for users to run them (they can download them and run them directly rather than copy/pasting) and it makes it at least hypothetically possible to write tests which import the code and run it.
  1. Don't say "Its suggested" in the docstrings.  First of all, it's "it's suggested", as in, "it is suggested".  Second, it's the passive voice: you, in the person of the omniscient API documentation, should just suggest the suggestion directly, and also give a reason for it.  So, something like: "If you're overriding `lookupProcedure`, you should also override `listProcedures`, so that the `system.listMethods` procedure will return an accurate view of available procedures to clients."

I think that's just about it; thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [31604]) Branching to 'xmlrpc-getfunction-4836'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [31605]) Apply issue_4836.patch 

refs [#4836](https://github.com/twisted/twisted/issues/4836)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [31607]) Rename getFunction to lookupProcedure; rename listFunctions to listProcedures; update tests and documentation to match.

refs [#4836](https://github.com/twisted/twisted/issues/4836)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [31608]) Add @since markers to new public APIs

refs [#4836](https://github.com/twisted/twisted/issues/4836)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [31609]) Finish renaming getFunction to lookupProcedure

refs [#4836](https://github.com/twisted/twisted/issues/4836)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [31610]) Add docstrings to TestLookupProcedure

refs [#4836](https://github.com/twisted/twisted/issues/4836)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [31611]) Reflow docstrings to 80ish columns

refs [#4836](https://github.com/twisted/twisted/issues/4836)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [31612]) Fix docstrings on the new test case

refs [#4836](https://github.com/twisted/twisted/issues/4836)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [31613]) Move the new XML-RPC example into a separate listing file

refs [#4836](https://github.com/twisted/twisted/issues/4836)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [31614]) Fix the suggestion in the lookupProcedure docstring

refs [#4836](https://github.com/twisted/twisted/issues/4836)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

I believe I covered all of the review points (changes linked above, one per point) except for point 6.  For that, I think it would be fine to file a new ticket to change the entire module to avoid doing real network I/O, rather than just implementing something for this one new test case.

Here are the [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/xmlrpc-getfunction-4836).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_19"></a>|@therve set owner to @exarkun|
|-|-|

```
+        raise xmlrpc.NoSuchFunction(
+            self.NOT_FOUND, 'procedure %s not found' % (procedureName,))
+
+
+class TestListProcedures(XMLRPC):
```

It's missing a blank line here.

Please merge!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_20"></a>|@therve commented|
|-|-|

Oh, and the news fragment needs to be fixed as the method was renamed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun commented|
|-|-|

(In [31674]) Address review feedback

refs [#4836](https://github.com/twisted/twisted/issues/4836)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun set status to `closed`|
|-|-|

(In [31675]) Merge xmlrpc-getfunction-4836 

Author: psykidellic, exarkun
Reviewer: glyph, therve
Fixes: [#4836](https://github.com/twisted/twisted/issues/4836)

Add a public overridable API to `twisted.web.xmlrpc.XMLRPC` which allows
customization of XML-RPC procedure dispatch.
